src/main/java/com/mongodb/jdbc/BsonTypeInfo.java

```diff
@@ -1,342 +1,230 @@
-            case DATE_TYPE_NAME:
-                return 8;
-            case INT_TYPE_NAME:
-                return 4;
-            case BOOL_TYPE_NAME:
-                return 1;
-            case BINDATA_TYPE_NAME:
-            case STRING_TYPE_NAME:
-            case ARRAY_TYPE_NAME:
-            case OBJECT_TYPE_NAME:
-            case OBJECTID_TYPE_NAME:
-            case DBPOINTER_TYPE_NAME:
-            case JAVASCRIPT_TYPE_NAME:
-            case JAVASCRIPTWITHSCOPE_TYPE_NAME:
-            case MAXKEY_TYPE_NAME:
-            case MINKEY_TYPE_NAME:
-            case REGEX_TYPE_NAME:
-            case SYMBOL_TYPE_NAME:
-            case TIMESTAMP_TYPE_NAME:
-            case UNDEFINED_TYPE_NAME:
-            case NULL_TYPE_NAME:
-            case BSON_TYPE_NAME:
-                return null;
+    /**
+     * Converts a bson type name to a BsonTypeInfo
+     *
+     * @param typeName is the bson type string as returned from the mongodb aggregation $type
+     *     function, this list of viable inputs is in the BSON_TYPE_NAMES static Set above.
+     * @return BsonTypeInfo object corresponding to bson type name.
+     */
+    public static BsonTypeInfo getBsonTypeInfoByName(String typeName) throws SQLException {
+        if (!BSON_TYPE_NAMES.contains(typeName)) {
+            throw new SQLException("Unknown bson type name: \"" + typeName + "\"");
         }
-        throw new SQLException("unknown bson typeName: " + typeName);
-    }
 
-    public static int getJDBCType(String typeName) throws SQLException {
-        switch (typeName) {
-            case ARRAY_TYPE_NAME:
-            case DBPOINTER_TYPE_NAME:
-            case OBJECT_TYPE_NAME:
-            case JAVASCRIPT_TYPE_NAME:
-            case JAVASCRIPTWITHSCOPE_TYPE_NAME:
-            case MAXKEY_TYPE_NAME:
-            case MINKEY_TYPE_NAME:
-            case OBJECTID_TYPE_NAME:
-            case REGEX_TYPE_NAME:
-            case SYMBOL_TYPE_NAME:
-            case TIMESTAMP_TYPE_NAME:
-            case UNDEFINED_TYPE_NAME:
-                return Types.OTHER;
-            case BINDATA_TYPE_NAME:
-                return Types.BINARY;
-            case BOOL_TYPE_NAME:
-                return Types.BIT;
-            case DATE_TYPE_NAME:
-                return Types.TIMESTAMP;
-            case INT_TYPE_NAME:
-                return Types.INTEGER;
-            case LONG_TYPE_NAME:
-                return Types.BIGINT;
-            case NULL_TYPE_NAME:
-                return Types.NULL;
-            case STRING_TYPE_NAME:
-                return Types.LONGVARCHAR;
-            case DECIMAL_TYPE_NAME:
-                return Types.DECIMAL;
-            case DOUBLE_TYPE_NAME:
-                return Types.DOUBLE;
+        // All type names can be determined uniquely off a combination of first letter
+        // and length except for "minKey" vs "maxKey" and "string" vs "symbol".
+        switch (typeName.charAt(0)) {
+            case 'a':
+                return BSON_ARRAY;
```

Comment:
Just a note that this method might be a source of errors later on.
I suppose this is what we are going for this because it should be faster than a switch, correct?

If somehow I end up passing the typeName 'Apricot' then it will return BSON_ARRAY. I assume that since we control the calls to the method, it should not be an issue, but when adding new types, it can be easy to forget to update this function and then the new type might end up as an unexpected BSON type or a SQLException.

I am fine to keep it as is if you considered this and think that our tests will protect us well enough and this should not happen.

Reply:
This method was just moved from one file to this one. It is meant to be used only with names provided in ADL collection schemas. Since those schemas are json schema validators, it is impossible for them to have invalid names; if they had invalid type names, ADL would have failed to create them (I suppose technically ADL could be buggy, but I don't think we should write code that handles ADL returning invalid information).

> I suppose this is what we are going for this because it should be faster than a switch, correct?

Yes, this is for speed! This is a common pattern I've seen in a couple repos on different teams that deal with bson parsing.

> but when adding new types, it can be easy to forget to update this function and then the new type might end up as an unexpected BSON type or a SQLException.

This is true in theory, but the introduction of a new BSON type would certainly result in a downstream ticket notification so we'd be aware of it before it is released and we'd be able to update this method appropriately.

Reply:
Sounds good to me 👍